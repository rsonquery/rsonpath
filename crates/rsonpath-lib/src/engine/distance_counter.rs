use lazy_static::lazy_static;
use log::debug;
use std::collections::HashMap;
use std::fs::File;
use std::io::{BufWriter, Write};
use std::path::Path;
use std::sync::Mutex;

// Track how many times every distance appears in the skipping process, only track the ones with frequency >= 1
lazy_static! {
    static ref DISTANCE_COUNTER_MAP: Mutex<HashMap<usize, usize>> = Mutex::new(HashMap::new());
}

// Increment the value by 1, or initialize to 1 if not present
pub fn increment_value(distance: usize) {
    let mut map = DISTANCE_COUNTER_MAP.lock().unwrap();
    *map.entry(distance).or_insert(0) += 1;
}

// Save HashMap to a CSV file
pub fn save_to_csv(file_path: &str) -> std::io::Result<()> {
    debug!("Saving to {}", file_path);
    let map = DISTANCE_COUNTER_MAP.lock().unwrap();
    let path = Path::new(file_path);
    let file = File::create(path)?;
    let mut writer = BufWriter::new(file);

    for (key, value) in map.iter() {
        debug!("Writing {}, {}", key, value);
        writeln!(writer, "{},{}", key, value)?;
    }

    Ok(())
}
