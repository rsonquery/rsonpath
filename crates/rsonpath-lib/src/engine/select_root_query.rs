//! Special case handlers for the empty query $.
//!
//! The main engine is built with a path automaton in mind, and for simplicity
//! we assume the root opening was already read. This makes it incompatible with
//! an empty query. Instead of rewriting the engine we provide fast-path implementations
//! here.
use crate::{
    engine::{error::EngineError, Input},
    input::{error::InputErrorConvertible, InputBlockIterator},
    is_json_whitespace,
    result::{empty::EmptyRecorder, Match, MatchCount, MatchIndex, MatchSpan, Sink},
    BLOCK_SIZE,
};
use crate::result::InputRecorder;

/// Count for an empty query &ndash; determine if the root exists.
pub(super) fn count<'i, 'r, I, R, const N: usize>(input: &I) -> Result<MatchCount, EngineError>
where
    I: Input<'i, 'r, R, N>,
    R: InputRecorder<I::Block> + 'r
{
    // Assuming a correct JSON, there is either one root if any non-whitespace character
    // occurs in the document, or the document is empty.
    if input.seek_non_whitespace_forward(0).e()?.is_some() {
        Ok(1)
    } else {
        Ok(0)
    }
}

/// Index for an empty query &ndash; determine the first index of the root.
pub(super) fn index<'i, 'r, I, R, S, const N: usize>(input: &I, sink: &mut S) -> Result<(), EngineError>
where
    I: Input<'i, 'r, R, N>,
    R: InputRecorder<I::Block> + 'r,
    S: Sink<MatchIndex>,
{
    // Assuming a correct JSON, the root starts at the first non-whitespace character, if any.
    if let Some((first_idx, _)) = input.seek_non_whitespace_forward(0).e()? {
        sink.add_match(first_idx - input.leading_padding_len())
            .map_err(|err| EngineError::SinkError(Box::new(err)))?;
    }

    Ok(())
}

/// Approximate span for an empty query &ndash; determine the first index and the length of the root.
pub(super) fn approx_span<'i, 'r, I, R, S, const N: usize>(input: &I, sink: &mut S) -> Result<(), EngineError>
where
    I: Input<'i, 'r, R, N>,
    R: InputRecorder<I::Block> + 'r,
    S: Sink<MatchSpan>,
{
    // The root spans the entire document, by definition, with the exception of whitespace.
    // We need to find the start index exactly, and then can return the length of the rest as the approximate
    // length of the root.
    //
    // Some input know their lengths: bytes already in memory, file mmaps, etc.
    // A BufferedInput over an arbitrary Read stream cannot know its length, so we actually
    // need to iterate until the end and count the bytes.
    if let Some((first_idx, _)) = input.seek_non_whitespace_forward(0).e()? {
        let end_idx = match input.len_hint() {
            Some(end_idx) => end_idx, // Known length, just take it.
            None => {
                // Unknown length, iterate and count.
                let mut iter = input.iter_blocks(&EmptyRecorder);
                let mut end_idx = 0;

                while (iter.next().e()?).is_some() {
                    end_idx += BLOCK_SIZE;
                }

                end_idx
            }
        };

        sink.add_match(MatchSpan::from_indices(
            first_idx - input.leading_padding_len(),
            end_idx - input.leading_padding_len(),
        ))
        .map_err(|err| EngineError::SinkError(Box::new(err)))?;
    }

    Ok(())
}

/// Match for an empty query &ndash; copy the entire document, trimming whitespace.
pub(super) fn match_<'i, 'r, I, R, S, const N: usize>(input: &I, sink: &mut S) -> Result<(), EngineError>
where
    I: Input<'i, 'r, R, N>,
    R: InputRecorder<I::Block> + 'r,
    S: Sink<Match>,
{
    // For a full match we need to copy the entire input starting from first non-whitespace,
    // and then trim the whitespace from the end. This might be slow if the document is excessively
    // padded with whitespace at start and/or end, but that's a pathological case.
    let mut iter = input.iter_blocks(&EmptyRecorder);
    let mut res: Vec<u8> = vec![];
    let mut first_significant_idx = None;
    let mut offset = 0;

    while let Some(block) = iter.next().e()? {
        if first_significant_idx.is_none() {
            // Start of the root not found yet, look for it.
            first_significant_idx = block.iter().position(|&x| !is_json_whitespace(x));

            if let Some(first_idx) = first_significant_idx {
                // Start of the root found in this block, copy the relevant part.
                res.extend(&block[first_idx..]);
            } else {
                offset += block.len();
            }
        } else {
            // Start of the root was already found, now we are copying everything.
            res.extend(&*block);
        }
    }

    if let Some(start) = first_significant_idx {
        // Trim whitespace if we have a result.
        while !res.is_empty() && is_json_whitespace(res[res.len() - 1]) {
            res.pop();
        }

        let actual_start = start + offset - input.leading_padding_len();
        sink.add_match(Match::from_start_and_bytes(actual_start, res))
            .map_err(|err| EngineError::SinkError(Box::new(err)))?;
    }

    Ok(())
}
