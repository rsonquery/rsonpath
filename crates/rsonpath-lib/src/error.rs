//! Common errors shared across the library.
use std::fmt::{self, Display};
use thiserror::Error;

pub(crate) const FEATURE_REQUEST_URL: &str =
    "https://github.com/V0ldek/rsonpath/issues/new?template=feature_request.md";
pub(crate) const BUG_REPORT_URL: &str = "https://github.com/V0ldek/rsonpath/issues/new?template=bug_report.md";

/// Internal irrecoverable error. These are caused solely
/// by bugs &ndash; broken invariants or assertions in internal logic &ndash;
/// but we return those instead of panicking.
#[derive(Error, Debug)]
pub struct InternalRsonpathError {
    details: &'static str,
    #[source]
    source: Option<InternalErrorSource>,
}

/// Errors in internal depth tracking of execution engines.
#[derive(Error, Debug)]
pub enum DepthError {
    /// The engine's maximum depth limit was exceeded.
    /// The inner [`usize`] indicates that limit.
    #[error("Maximum depth of {0} exceeded.")]
    AboveLimit(usize),
    /// The document has unmatched closing characters
    /// and is malformed.
    #[error("Depth fell below zero.")]
    BelowZero,
}

struct InternalErrorSource(Box<dyn std::error::Error + Send + Sync>);

impl fmt::Debug for InternalErrorSource {
    #[inline(always)]
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(&self.0, f)
    }
}

impl Display for InternalErrorSource {
    #[inline(always)]
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Display::fmt(&self.0, f)
    }
}

impl std::error::Error for InternalErrorSource {
    #[inline(always)]
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        self.0.source()
    }
}

impl InternalRsonpathError {
    #[allow(unused)]
    pub(crate) fn from_expectation(details: &'static str) -> Self {
        Self { details, source: None }
    }

    #[allow(unused)]
    pub(crate) fn from_error<E: std::error::Error + Send + Sync + 'static>(err: E, details: &'static str) -> Self {
        Self {
            details,
            source: Some(InternalErrorSource(Box::new(err))),
        }
    }
}

impl Display for InternalRsonpathError {
    #[inline]
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let details_are_enabled = std::env::var("RUST_BACKTRACE").unwrap_or_default() == "0";
        write!(
            f,
            "an internal error has occurred; this is a bug, please report it at {BUG_REPORT_URL}"
        )?;

        if details_are_enabled {
            writeln!(f, "; the error details follow")?;
            write!(f, "{}", self.details)?;
            if let Some(source) = &self.source {
                write!(f, "; source: {source}")?;
            }
        }

        Ok(())
    }
}

/// Error raised when rsonpath is asked to perform an operation that is currently
/// unsupported. This may be either because the feature is in the works, or
/// because it is not planned to ever be supported.
#[derive(Error, Debug)]
pub struct UnsupportedFeatureError {
    issue: Option<usize>,
    feature: &'static str,
}

impl UnsupportedFeatureError {
    #[must_use]
    #[allow(dead_code)]
    #[inline(always)]
    fn tracked(issue: usize, feature: &'static str) -> Self {
        Self {
            issue: Some(issue),
            feature,
        }
    }

    #[must_use]
    #[inline(always)]
    fn untracked(feature: &'static str) -> Self {
        Self { issue: None, feature }
    }

    /// Large JSON Depths feature &ndash; supporting JSON documents
    /// with nesting depth exceeding 255. Unsupported and not planned.
    #[must_use]
    #[inline(always)]
    pub fn large_json_depths() -> Self {
        Self::untracked("Large JSON Depths")
    }

    /// Large Automaton Queries feature &ndash; supporting queries that
    /// cause compiled DFAs to exceed 256 states. Unsupported and not planned.
    #[must_use]
    #[inline(always)]
    pub fn large_automaton_queries() -> Self {
        Self::untracked("Large Automaton Queries")
    }

    /// Multiple Selector per Segment feature &ndash; supporting queries
    /// that contain a union of selectors per step. Unsupported and not planned (yet).
    #[must_use]
    #[inline(always)]
    pub fn multiple_selectors() -> Self {
        Self::untracked("Multiple Selector per Segment")
    }

    /// Indexing from End &ndash; supporting index and slice selectors that
    /// use from-end indexing. Unsupported and not planned (yet).
    #[must_use]
    #[inline(always)]
    pub fn indexing_from_end() -> Self {
        Self::untracked("Indexing from End")
    }

    /// Slice Selector &ndash; supporting slice selectors.
    /// <https://github.com/V0ldek/rsonpath/issues/152>
    #[must_use]
    #[inline(always)]
    pub fn slice_selector() -> Self {
        Self::tracked(152, "Slice Selector")
    }

    /// Slice with Backward Step &ndash; supporting slice selectors that step backwards.
    /// Unsupported and not planned (yet).
    #[must_use]
    #[inline(always)]
    pub fn slice_with_backward_step() -> Self {
        Self::untracked("Slice with Backward Step")
    }

    /// Filter Selector &ndash; supporting filter selectors.
    /// <https://github.com/V0ldek/rsonpath/issues/154>
    #[must_use]
    #[inline(always)]
    pub fn filter_selector() -> Self {
        Self::tracked(154, "Filter Selector")
    }

    /// Returns the issue number on GitHub corresponding to the unsupported feature.
    /// Is [`None`] if the feature is not planned.
    #[must_use]
    #[inline(always)]
    pub fn issue(&self) -> Option<usize> {
        self.issue
    }

    /// Returns the descriptive name of the feature.
    #[must_use]
    #[inline(always)]
    pub fn feature(&self) -> &str {
        self.feature
    }

    /// Whether the issue is planned to ever be supported.
    #[must_use]
    #[inline(always)]
    pub fn is_planned(&self) -> bool {
        self.issue.is_some()
    }
}

impl Display for UnsupportedFeatureError {
    #[inline]
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self.issue {
            Some(issue) => {
                write!(
                    f,
                    "the feature {} (#{}) is not supported yet; it is being tracked and discussed at \
                    https://github.com/V0ldek/rsonpath/issues/{}",
                    self.feature, issue, issue
                )
            }
            None => {
                write!(
                    f,
                    "the feature {} is not supported, and is not planned; \
                    if you would like to see it introduced to rsonpath, please raise a feature request at \
                    {FEATURE_REQUEST_URL}",
                    self.feature
                )
            }
        }
    }
}
