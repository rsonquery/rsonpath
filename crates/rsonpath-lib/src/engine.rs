//! Base traits for different implementations of JSONPath execution engines.
//!
//! Defines the [`Engine`] trait that provides different ways of retrieving
//! query results from input bytes, as well as [`Compiler`] which provides
//! a standalone entry point for compiling a [`JsonPathQuery`] into an [`Engine`].
pub mod error;
mod head_skipping;
pub mod main;
mod tail_skipping;
pub use main::MainEngine as RsonpathEngine;
mod select_root_query;

use self::error::EngineError;
use crate::input::SeekableBackwardsInput;
use crate::{
    automaton::{error::CompilerError, Automaton},
    input::Input,
    result::{Match, MatchCount, MatchIndex, MatchSpan, Sink},
};
use rsonpath_syntax::JsonPathQuery;

/// An engine that can run its query on a given input.
pub trait Engine {
    /// Find the number of matches on the given [`Input`].
    ///
    /// The result is equivalent to using [`matches`](Engine::matches) and counting the matches,
    /// but in general is much more time and memory efficient.
    ///
    /// # Errors
    /// An appropriate [`EngineError`] is returned if the JSON input is malformed
    /// and the syntax error is detected.
    ///
    /// **Please note** that detecting malformed JSONs is not guaranteed.
    /// Some glaring errors like mismatched braces or double quotes are raised,
    /// but in general **the result of an engine run on an invalid JSON is undefined**.
    /// It _is_ guaranteed that the computation terminates and does not panic.
    fn count<I>(&self, input: &I) -> Result<MatchCount, EngineError>
    where
        I: SeekableBackwardsInput;

    /// Find the starting indices of matches on the given [`Input`] and write them to the [`Sink`].
    ///
    /// The result is equivalent to using [`matches`](Engine::matches) and extracting the
    /// [`Match::span.start_idx`],
    /// but in general is much more time and memory efficient.
    ///
    /// # Errors
    /// An appropriate [`EngineError`] is returned if the JSON input is malformed
    /// and the syntax error is detected.
    ///
    /// **Please note** that detecting malformed JSONs is not guaranteed.
    /// Some glaring errors like mismatched braces or double quotes are raised,
    /// but in general **the result of an engine run on an invalid JSON is undefined**.
    /// It _is_ guaranteed that the computation terminates and does not panic.
    fn indices<I, S>(&self, input: &I, sink: &mut S) -> Result<(), EngineError>
    where
        I: SeekableBackwardsInput,
        S: Sink<MatchIndex>;

    /// Find the approximate spans of matches on the given [`Input`] and write them to the [`Sink`].
    ///
    /// "Approximate" means that the ends of spans are not guaranteed to be located exactly at the end of a match,
    /// but may include trailing whitespace. **Importantly**, it may be beyond the rigidly-defined length of the
    /// input, as the engine is allowed to pad the input with whitespace to help with processing. For the purposes
    /// of this API, it is assumed that every character after the logical end of the input is a whitespace character.
    /// With that in mind, it is guaranteed that:
    /// 1. the span start is exact;
    /// 2. the span encompasses the entire matched value;
    /// 3. the only characters included after the value are JSON whitespace characters:
    ///    space (0x20), horizontal tab (0x09), new line (0x0A), carriage return (0x0D).
    ///
    /// # Errors
    /// An appropriate [`EngineError`] is returned if the JSON input is malformed
    /// and the syntax error is detected.
    ///
    /// **Please note** that detecting malformed JSONs is not guaranteed.
    /// Some glaring errors like mismatched braces or double quotes are raised,
    /// but in general **the result of an engine run on an invalid JSON is undefined**.
    /// It _is_ guaranteed that the computation terminates and does not panic.
    fn approximate_spans<I, S>(&self, input: &I, sink: &mut S) -> Result<(), EngineError>
    where
        I: SeekableBackwardsInput,
        S: Sink<MatchSpan>;

    /// Find all matches on the given [`Input`] and write them to the [`Sink`].
    ///
    /// # Errors
    /// An appropriate [`EngineError`] is returned if the JSON input is malformed
    /// and the syntax error is detected.
    ///
    /// **Please note** that detecting malformed JSONs is not guaranteed.
    /// Some glaring errors like mismatched braces or double quotes are raised,
    /// but in general **the result of an engine run on an invalid JSON is undefined**.
    /// It _is_ guaranteed that the computation terminates and does not panic.
    fn matches<I, S>(&self, input: &I, sink: &mut S) -> Result<(), EngineError>
    where
        I: SeekableBackwardsInput,
        S: Sink<Match>;
}

/// An engine that can be created by compiling a [`JsonPathQuery`].
pub trait Compiler {
    /// Concrete type of the [`Engines`](`Engine`) created,
    /// parameterized with the lifetime of the input query.
    type E<'q>: Engine + 'q;

    /// Compile a [`JsonPathQuery`] into an [`Engine`].c
    ///
    /// # Errors
    /// An appropriate [`CompilerError`] is returned if the compiler
    /// cannot handle the query.
    fn compile_query(query: &JsonPathQuery) -> Result<Self::E<'_>, CompilerError>;

    /// Turn a compiled [`Automaton`] into an [`Engine`].
    fn from_compiled_query(automaton: Automaton<'_>) -> Self::E<'_>;
}
