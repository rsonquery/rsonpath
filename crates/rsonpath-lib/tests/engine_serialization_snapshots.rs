#[cfg(feature = "serde")]
mod ron {
    use insta::assert_ron_snapshot;
    use rsonpath::engine::{Compiler, RsonpathEngine};
    use std::error::Error;

    fn engine(string: &str) -> Result<RsonpathEngine, Box<dyn Error>> {
        let query = rsonpath_syntax::parse(string)?;
        let engine = RsonpathEngine::compile_query(&query)?;
        Ok(engine)
    }

    #[test]
    fn empty_query() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&engine("$")?);
        Ok(())
    }

    #[test]
    #[cfg(target_pointer_width = "64")]
    fn readme_query_64() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&engine("$.jsonpath[*]")?);
        Ok(())
    }

    #[test]
    #[cfg(target_pointer_width = "32")]
    fn readme_query_32() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&engine("$.jsonpath[*]")?);
        Ok(())
    }

    #[test]
    #[cfg(target_pointer_width = "64")]
    fn jsonpath_example_query_64() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&engine("$..phoneNumbers[*].number")?);
        Ok(())
    }

    #[test]
    #[cfg(target_pointer_width = "32")]
    fn jsonpath_example_query_32() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&engine("$..phoneNumbers[*].number")?);
        Ok(())
    }

    #[test]
    #[cfg(target_pointer_width = "64")]
    fn real_life_query_64() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&engine("$.personal.details.contact.information.phones.home")?);
        Ok(())
    }

    #[test]
    #[cfg(target_pointer_width = "32")]
    fn real_life_query_32() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&engine("$.personal.details.contact.information.phones.home")?);
        Ok(())
    }

    #[test]
    #[cfg(target_pointer_width = "64")]
    fn slice_64() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&engine("$..entries[3:5:7]")?);
        Ok(())
    }

    #[test]
    #[cfg(target_pointer_width = "32")]
    fn slice_32() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&engine("$..entries[3:5:7]")?);
        Ok(())
    }
}
