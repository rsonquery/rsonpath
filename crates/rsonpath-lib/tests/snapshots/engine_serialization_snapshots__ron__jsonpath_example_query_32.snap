---
source: crates/rsonpath-lib/tests/engine_serialization_snapshots.rs
expression: "&engine(\"$..phoneNumbers[*].number\")?"
---
(V1, Automaton(
  states: [
    StateTable(
      attributes: StateAttributes(2),
      member_transitions: [],
      array_transitions: [],
      fallback_state: State(0),
    ),
    StateTable(
      attributes: StateAttributes(0),
      member_transitions: [
        (StringPattern(
          bytes: [
            34,
            112,
            104,
            111,
            110,
            101,
            78,
            117,
            109,
            98,
            101,
            114,
            115,
            34,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
          ],
          alternatives: [
            None,
            SlashUSingle(808923184, 1),
            SlashUSingle(943075376, 1),
            SlashUSingle(1714827312, 1),
            SlashUSingle(1698050096, 1),
            SlashUSingle(892743728, 1),
            SlashUSingle(1697919024, 1),
            SlashUSingle(892809264, 1),
            SlashUSingle(1681272880, 1),
            SlashUSingle(842412080, 1),
            SlashUSingle(892743728, 1),
            SlashUSingle(842477616, 1),
            SlashUSingle(859254832, 1),
            None,
          ],
          len: 14,
          len_limit: 74,
        ), State(2)),
      ],
      array_transitions: [],
      fallback_state: State(1),
    ),
    StateTable(
      attributes: StateAttributes(0),
      member_transitions: [
        (StringPattern(
          bytes: [
            34,
            112,
            104,
            111,
            110,
            101,
            78,
            117,
            109,
            98,
            101,
            114,
            115,
            34,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
          ],
          alternatives: [
            None,
            SlashUSingle(808923184, 1),
            SlashUSingle(943075376, 1),
            SlashUSingle(1714827312, 1),
            SlashUSingle(1698050096, 1),
            SlashUSingle(892743728, 1),
            SlashUSingle(1697919024, 1),
            SlashUSingle(892809264, 1),
            SlashUSingle(1681272880, 1),
            SlashUSingle(842412080, 1),
            SlashUSingle(892743728, 1),
            SlashUSingle(842477616, 1),
            SlashUSingle(859254832, 1),
            None,
          ],
          len: 14,
          len_limit: 74,
        ), State(4)),
      ],
      array_transitions: [],
      fallback_state: State(3),
    ),
    StateTable(
      attributes: StateAttributes(8),
      member_transitions: [
        (StringPattern(
          bytes: [
            34,
            112,
            104,
            111,
            110,
            101,
            78,
            117,
            109,
            98,
            101,
            114,
            115,
            34,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
          ],
          alternatives: [
            None,
            SlashUSingle(808923184, 1),
            SlashUSingle(943075376, 1),
            SlashUSingle(1714827312, 1),
            SlashUSingle(1698050096, 1),
            SlashUSingle(892743728, 1),
            SlashUSingle(1697919024, 1),
            SlashUSingle(892809264, 1),
            SlashUSingle(1681272880, 1),
            SlashUSingle(842412080, 1),
            SlashUSingle(892743728, 1),
            SlashUSingle(842477616, 1),
            SlashUSingle(859254832, 1),
            None,
          ],
          len: 14,
          len_limit: 74,
        ), State(2)),
        (StringPattern(
          bytes: [
            34,
            110,
            117,
            109,
            98,
            101,
            114,
            34,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
          ],
          alternatives: [
            None,
            SlashUSingle(1698050096, 1),
            SlashUSingle(892809264, 1),
            SlashUSingle(1681272880, 1),
            SlashUSingle(842412080, 1),
            SlashUSingle(892743728, 1),
            SlashUSingle(842477616, 1),
            None,
          ],
          len: 8,
          len_limit: 38,
        ), State(6)),
      ],
      array_transitions: [],
      fallback_state: State(1),
    ),
    StateTable(
      attributes: StateAttributes(8),
      member_transitions: [
        (StringPattern(
          bytes: [
            34,
            112,
            104,
            111,
            110,
            101,
            78,
            117,
            109,
            98,
            101,
            114,
            115,
            34,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
          ],
          alternatives: [
            None,
            SlashUSingle(808923184, 1),
            SlashUSingle(943075376, 1),
            SlashUSingle(1714827312, 1),
            SlashUSingle(1698050096, 1),
            SlashUSingle(892743728, 1),
            SlashUSingle(1697919024, 1),
            SlashUSingle(892809264, 1),
            SlashUSingle(1681272880, 1),
            SlashUSingle(842412080, 1),
            SlashUSingle(892743728, 1),
            SlashUSingle(842477616, 1),
            SlashUSingle(859254832, 1),
            None,
          ],
          len: 14,
          len_limit: 74,
        ), State(4)),
        (StringPattern(
          bytes: [
            34,
            110,
            117,
            109,
            98,
            101,
            114,
            34,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
          ],
          alternatives: [
            None,
            SlashUSingle(1698050096, 1),
            SlashUSingle(892809264, 1),
            SlashUSingle(1681272880, 1),
            SlashUSingle(842412080, 1),
            SlashUSingle(892743728, 1),
            SlashUSingle(842477616, 1),
            None,
          ],
          len: 8,
          len_limit: 38,
        ), State(5)),
      ],
      array_transitions: [],
      fallback_state: State(3),
    ),
    StateTable(
      attributes: StateAttributes(9),
      member_transitions: [
        (StringPattern(
          bytes: [
            34,
            112,
            104,
            111,
            110,
            101,
            78,
            117,
            109,
            98,
            101,
            114,
            115,
            34,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
          ],
          alternatives: [
            None,
            SlashUSingle(808923184, 1),
            SlashUSingle(943075376, 1),
            SlashUSingle(1714827312, 1),
            SlashUSingle(1698050096, 1),
            SlashUSingle(892743728, 1),
            SlashUSingle(1697919024, 1),
            SlashUSingle(892809264, 1),
            SlashUSingle(1681272880, 1),
            SlashUSingle(842412080, 1),
            SlashUSingle(892743728, 1),
            SlashUSingle(842477616, 1),
            SlashUSingle(859254832, 1),
            None,
          ],
          len: 14,
          len_limit: 74,
        ), State(2)),
        (StringPattern(
          bytes: [
            34,
            110,
            117,
            109,
            98,
            101,
            114,
            34,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
          ],
          alternatives: [
            None,
            SlashUSingle(1698050096, 1),
            SlashUSingle(892809264, 1),
            SlashUSingle(1681272880, 1),
            SlashUSingle(842412080, 1),
            SlashUSingle(892743728, 1),
            SlashUSingle(842477616, 1),
            None,
          ],
          len: 8,
          len_limit: 38,
        ), State(6)),
      ],
      array_transitions: [],
      fallback_state: State(1),
    ),
    StateTable(
      attributes: StateAttributes(1),
      member_transitions: [
        (StringPattern(
          bytes: [
            34,
            112,
            104,
            111,
            110,
            101,
            78,
            117,
            109,
            98,
            101,
            114,
            115,
            34,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
          ],
          alternatives: [
            None,
            SlashUSingle(808923184, 1),
            SlashUSingle(943075376, 1),
            SlashUSingle(1714827312, 1),
            SlashUSingle(1698050096, 1),
            SlashUSingle(892743728, 1),
            SlashUSingle(1697919024, 1),
            SlashUSingle(892809264, 1),
            SlashUSingle(1681272880, 1),
            SlashUSingle(842412080, 1),
            SlashUSingle(892743728, 1),
            SlashUSingle(842477616, 1),
            SlashUSingle(859254832, 1),
            None,
          ],
          len: 14,
          len_limit: 74,
        ), State(2)),
      ],
      array_transitions: [],
      fallback_state: State(1),
    ),
  ],
))
